python/tests/test_schema.py

```diff
@@ -55,270 +60,194 @@ def test_table_schema_pyarrow_020():
-            "nullable": False,
-            "type": {"name": "dictionary"},
-            "dictionary": {"indexType": {"type": {"name": "int", "bitWidth": 8}}},
-            "children": [{"type": {"name": "int", "bitWidth": 32}}],
-        }
-    )
-    assert pyarrow_field.name == field_name
-    assert pyarrow_field.type == pyarrow.map_(pyarrow.int8(), pyarrow.int32())
-    assert pyarrow_field.metadata == metadata
-    assert pyarrow_field.nullable is False
-
-    pyarrow_field = pyarrow_field_from_dict(
-        {
-            "name": field_name,
-            "type": {"name": "struct"},
-            "children": [
-                {
-                    "name": "x",
-                    "type": {"name": "int", "bitWidth": 64},
-                    "nullable": True,
-                    "metadata": {},
-                }
-            ],
-            "metadata": {"metadata_k": "metadata_v"},
-            "nullable": False,
-        }
-    )
-    assert pyarrow_field.name == field_name
-    assert pyarrow_field.type == pyarrow.struct(
-        [pyarrow.field("x", pyarrow.int64(), True, {})]
-    )
-    assert pyarrow_field.metadata == metadata
-    assert pyarrow_field.nullable is False
+    ]
+    for key_type, value_type, value_contains_null in init_values:
+        map_type = MapType(key_type, value_type, value_contains_null)
+
+        assert map_type.type == "map"
+        assert map_type.key_type == key_type
+        assert map_type.value_type == value_type
+        assert map_type.value_contains_null == value_contains_null
+
+        # Map type is not yet supported in C Data Interface
+        # https://github.com/apache/arrow-rs/issues/2037
+        # pa_type = map_type.to_pyarrow()
+        # assert map_type == PrimitiveType.from_pyarrow(pa_type)
+
+        json_type = map_type.to_json()
+        assert map_type == MapType.from_json(json_type)
+
+
+def test_struct_delta_types():
+    fields = [
+        Field("x", "integer", nullable=True, metadata={"x": {"y": 3}}),
+        Field("y", PrimitiveType("string"), nullable=False),
+    ]
+
+    struct_type = StructType(fields)
+
+    assert struct_type.type == "struct"
+    assert struct_type.fields == fields
+
+    json_type = struct_type.to_json()
+    assert struct_type == StructType.from_json(json_type)
+
+    # Field metadata doesn't roundtrip currently
+    # See: https://github.com/apache/arrow-rs/issues/478
+    fields = [
+        Field("x", "integer", nullable=True),
+        Field("y", PrimitiveType("string"), nullable=False),
+    ]
+    struct_type = StructType(fields)
+    pa_type = struct_type.to_pyarrow()
+    assert struct_type == StructType.from_pyarrow(pa_type)
+
+
+def test_delta_field():
+    args = [
+        ("x", PrimitiveType("string"), True, {}),
+        ("y", "float", False, {"x": {"y": 3}}),
+        ("z", ArrayType(StructType([Field("x", "integer", True)]), True), True, None),
+    ]
+
+    # TODO: are there field names we should reject?
```

Comment:
the only thing i can think of is when we use the field names for partitioning. We have some test for special characters. Not sure though if we could handle (or should be able to) if an "=" were to appear in the field name.

Reply:
I think I may leave this for later. There may be complications with column mapping.